operator/controllers/resources/reconciler.go

```diff
@@ -0,0 +1,142 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resources
+
+import (
+	"context"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	ctrl "sigs.k8s.io/controller-runtime"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+)
+
+// DeletionFinalizer is a name of finalizer added to resource's 'finalizers' field
+// for tracking deletion events.
+const DeletionFinalizer = "resources.teleport.dev/deletion"
+
+type DeleteExternal func(context.Context, kclient.Object) error
+type UpsertExternal func(context.Context, kclient.Object) error
+
+type ResourceBaseReconciler struct {
+	kclient.Client
+	DeleteExternal DeleteExternal
+	UpsertExternal UpsertExternal
+}
+
+/*
+Do will receive an update request and reconcile the resource.
+
+When an event arrives we must propagate that change into the Teleport cluster.
+We have two types of events: update/create and delete.
+
+For creating/updating we check if the resource exists in Teleport
+- if it does, we update it
+- otherwise we create it
+Always using the state of the resource in the cluster as the source of truth.
+
+For deleting, the recommendation is to use finalizers.
+Finalizers allow us to map an external resource to a kubernetes resource.
+So, when we create or update a resource, we add our own finalizer to the kubernetes resource list of finalizers.
+
+For a delete event which has our finalizer: the resource is deleted in Teleport.
+If it doesn't have the finalizer, we do nothing.
+
+----
+
+Every time we update a resource in Kubernetes (adding finalizers or the OriginLabel), we end the reconciliation process.
+Afterwards, we receive the request again and we progress to the next step.
+This allow us to progress with smaller changes and avoid a long-running reconciliation.
+*/
+func (r ResourceBaseReconciler) Do(ctx context.Context, req ctrl.Request, obj kclient.Object) (ctrl.Result, error) {
+	// https://sdk.operatorframework.io/docs/building-operators/golang/advanced-topics/#external-resources
+	log := log.FromContext(ctx).WithValues("namespacedname", req.NamespacedName)
+
+	if err := r.Get(ctx, req.NamespacedName, obj); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("not found")
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "failed to get resource")
+		return ctrl.Result{}, trace.Wrap(err)
+	}
+
+	hasDeletionFinalizer := controllerutil.ContainsFinalizer(obj, DeletionFinalizer)
+	isMarkedToBeDeleted := !obj.GetDeletionTimestamp().IsZero()
+
+	// Delete
+	if isMarkedToBeDeleted {
+		if hasDeletionFinalizer {
+			log.Info("deleting object in Teleport")
+			if err := r.DeleteExternal(ctx, obj); err != nil {
```

Comment:
Should this handle not found error, in the case where the resource was deleted directly through Teleport rather than the operator? Right now this leads to:

```
1.6565323828787718e+09	INFO	deleting object in Teleport	{"controller": "user", "controllerGroup": "resources.teleport.dev", "controllerKind": "User", "role": {"name":"myrole","namespace":"teleport-cluster"}, "namespace": "teleport-cluster", "name": "myrole", "reconcileID": "846a414c-d2b9-457a-85c7-61c12fe713c4", "namespacedname": "teleport-cluster/myrole"}
1.6565323828816507e+09	ERROR	Reconciler error	{"controller": "user", "controllerGroup": "resources.teleport.dev", "controllerKind": "User", "role": {"name":"myrole","namespace":"teleport-cluster"}, "namespace": "teleport-cluster", "name": "myrole", "reconcileID": "846a414c-d2b9-457a-85c7-61c12fe713c4", "error": "role \"myrole\" is not found"}
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.1/pkg/internal/controller/controller.go:234
```

Can you add a test case for this scenario as well?

Reply:
Oh, great point. It feels weird to have the operator not reconcile when something happens to one of its managed resources, this is a common pain point when reconciling things not in Kubernetes.

Ideally it should be doing a WATCH to get updates from resources in Teleport but this can be overly complex to do. Maybe a Tick every X minute looking for changes would be enough (but this could drastically increase DB usage). A janky fix workaround could also be reducing the resync period from its default 10 hours to something shorter like 1 hour.

Reply:
I committed a failing test and a fix for it
When trying to delete but it errors with "NotFound" we continue as if it was not an error

Regarding the sync from Teleport into Kubernetes' CR, we decided to not implement that on our first iteration
We only support from Kubernetes to Teleport

Sorry, this is not written anywhere, afaik

Reply:
In this case I suggest reducing the resync period (maybe have this configurable?), it will ensure we converge back way faster if a drift happens